avocado/utils/asset.py

```diff
@@ -158,6 +178,10 @@ def fetch(self):
             else:
                 raise UnsupportedProtocolError("Unsupported protocol"
                                                ": %s" % urlobj.scheme)
+
+            dirname = os.path.dirname(self.asset_file)
+            if not os.path.isdir(dirname):
```

Comment:
This will fail in case the `dirname` is an existing file (asset with the same name as hash of this url). So I see following scenarios:

1. it does not exists; create it
2. it exists and is a dir (previously downloaded another item from the same base url); proceed
3. it exists and is not a dir (other file, link, ...); we need to remove it and create a dir. Note that you should use `FileLock` while removing, creating_dir and downloading the file.

For those reasons I'd probably add support for `cache_relative_dir` to our downloaders so they can first download the asset file (in case of url) and only acquire the lock when moving it locally. (so file-downloader would lock both while creating the symlink, url would lock after the download when copying it.